### PR TITLE
Move test libraries into testlib components, mildly sanitize

### PR DESCRIPTION
**Pull request by michaelpj**
This does something I've wanted to do for ages, and pulls the "test library" code that we have into testlib components, and organizes them a little better. At the moment I just moved things, but I'm going to do another pass after this and try and make things saner and more consistent too.

This makes it a lot easier to reuse the test library stuff. For `plutus-core` we were just exporting it in the main library (ew), but for e.g. `plutus-tx` we weren't. Which was annoying, because I wanted to use some of the code there for writing some tests in `plutus-benchmark`.

Not including the test stuff in `plutus-core` is also a win for our dependency footprint: now we no longer force a transitive dependency on `tasty` on all our downstream users!

### CODE REVIEW

File: plutus-core/plutus-core.cabal [hunk 1]
```diff
@@ -326,6 +299,61 @@ library
         word-array -any,
         cardano-crypto-class -any
 
+-- could split this up if we split up the main library for UPLC/PLC/PIR
```
Comment (thealmarty) [inline plutus-core/plutus-core.cabal:302]: It may be good to split up the main library for UPLC/PLC/PIR......probably should be in a separate PR though.

**Review by thealmarty — APPROVED** [review]
Yes I like this. Please do this.

**Comment by michaelpj** [comment]
> It may be good to split up the main library for UPLC/PLC/PIR......probably should be in a separate PR though.

The main thing that's preventing me from doing this is that then we'd actually have to disentangle them! There's a lot of common code in `plutus-core` that's used by the others, for example. So it would require quite a bit of work.

**Review by kwxm — APPROVED** [review]
No complaints: this all looks perfectly sensible, and using internal libraries seems to be very helpful to keep things tidy and  reduce repetition.

One comment/question though.  When I saw `Test.Tasty.Extras` appearing in  a number of files I initially assumed it was from some external library, not something that we were providing ourselves (I even unsuccessfully consulted Google to see what was in it).  Is it common practice to name things like this?

**Review by bezirg — APPROVED** [review]

**Comment by bezirg** [comment]
> One comment/question though.  When I saw `Test.Tasty.Extras` appearing in  a number of files I initially assumed it was from some external library, not something that we were providing ourselves (I even unsuccessfully consulted Google to see what was in it).  Is it common practice to name things like this?

I don't think there is any convention agreed among the community. The whole naming of haskell module paths is controversial to begin with...

**Comment by michaelpj** [comment]
re: `Test.Tasty.Extras`, I was actually following our own convention! We have a lot of modules called `Foo.Extras` which contain our own little extra bit for the library Foo (e.g. `Data.Aeson.Extras`). Not sure whether it's a *great* convention, but I think it's okay? Should write it down, though.

**Comment by kwxm** [comment]
> re: `Test.Tasty.Extras`, I was actually following our own convention!

OK, I seem to have managed not to notice that!  There are [tons of libraries](https://hackage.haskell.org/packages/search?terms=extras) called *-extras on Hackage, and I think that's why I was confused.  I  suppose we could have called these things something like `OurExtras` or `PlcExtras`, but it would be a bit pointless to change that at this stage.

**Comment by michaelpj** [comment]
Okay, let's do this.
